### PR TITLE
Fix nullable warning with an enumerator local function in a lambda

### LINKED ISSUE #42396: nullable on nested inner functions gives weird CS8619 warning
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/content/problem/936394/nullable-on-nested-inner-functions-gives-weird-cs8.html)._

---
[regression] [worked-in:16.x public preview]
The following program gives the warning 
> CS8619    Nullability of reference types in value of type 'T' doesn't match target type 'List<T>'.

on the `x` of line 13:
```cs
foreach (T x in xs) yield return x;
```

```cs
using System;
using System.Collections.Generic;

namespace NullableTest {
    public class NestedList<T> {
        public NestedList(Func<IEnumerable<T>> generator) { }
    }

internal class ObjBuilder<T> where T : class {
        public static void Build() {
            new NestedList<List<T>>(() => {
                IEnumerable<T> Enumerate(IEnumerable<T> xs) {
                    foreach (T x in xs) yield return x;
                }

return new List<T>[0];
            });
        }
    }
}
```

All of the types are non-nullable. And comparing `T` and `List<T>` on nullability doesn't make sense anyway. (This is simplified from a larger codebase which can be provided upon request.)

---
#### Daniel Perelman [MSFT] on 3/2/2020, 03:25 PM: 

(private comment, text removed)
#### Visual Studio Feedback System on 3/2/2020, 04:20 PM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

#### Daniel Perelman [MSFT] on 3/6/2020, 11:21 AM: 

(private comment, text removed)

---
### Original Solutions
(no solutions)

### PR DESCRIPTION
**Pull request by agocke**
Fixes #42396

### CODE REVIEW

**Comment by agocke** [comment]
@jcouv could you take a look?

**Comment by agocke** [comment]
ping @jcouv

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 1]
```diff
@@ -46358,6 +46358,32 @@ void f()
                 );
         }
 
+        [Fact]
+        [WorkItem(42396, "https://github.com/dotnet/roslyn/issues/42396")]
+        public void LocalFunction_05()
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: I see that we also have logic involving `_delegateInvokeMethod` in `NullableWalker.TryGetReturnType()` (used in `VisitReturnStatement`). Could we add a similar test case for vanilla `return` statement in non-enumerable nested local function?

Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: Makes me wonder if we need `_delegateInvokeMethod` at all if it's not for such nested scenarios...

**Comment by agocke** [comment]
@jcouv Good catch. I've implemented a work-around for analyzing return types, but this method of dealing with lambdas seems broken to me. Filed https://github.com/dotnet/roslyn/issues/42482

I took a glance at the lambda implementation and wasn't sure what the Snapshot stuff was all about.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 1 above)
Comment (agocke) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: We shouldn't except that lambdas are analyzed in a separate walker.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 2]
```diff
@@ -8351,7 +8353,7 @@ public override BoundNode VisitYieldReturnStatement(BoundYieldReturnStatement no
             {
                 return null;
             }
-            var method = _delegateInvokeMethod ?? (MethodSymbol)CurrentSymbol;
+            var method = (MethodSymbol)CurrentSymbol;
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: I didn't quite understand why the handling of `return` and `yield return` are diverging. What's the rationale?

Comment (agocke) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: `Yield return` is illegal in a lambda. This code was always wrong.

Comment (jcouv) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Ah, makes sense. Thanks

**Comment by jcouv** [comment]
LGTM Thanks

**Comment by agocke** [comment]
@dotnet/roslyn-compiler for review

**Pull request merged**